test/test_custom_op_testing.py

```diff
@@ -0,0 +1,327 @@
+        with self.assertRaisesRegex(
+                RuntimeError,
+                'FooBackward returned an invalid gradient at index 0'):
+            operator_compile_check(f, (x,), {})
+
+    def test_missing_functionalization(self, device):
+        lib = self.lib()
+        lib.define("foo(Tensor(a!) x) -> Tensor(a!)")
+        op = self.ns().foo.default
+
+        class Foo(torch.autograd.Function):
+            @staticmethod
+            def forward(ctx, x):
+                ctx.mark_dirty(x)
+                with torch._C._AutoDispatchBelowAutograd():
+                    return op(x)
+
+            @staticmethod
+            def backward(ctx, gx):
+                return gx
+
+        def foo_impl(x):
+            return x.sin_()
+
+        def foo_meta(x):
+            return x
+
+        lib.impl("foo", Foo.apply, "Autograd")
+        lib.impl("foo", foo_impl, "CPU")
+        lib.impl("foo", foo_impl, "CUDA")
+        lib.impl("foo", foo_meta, "Meta")
+
+        def f(x):
+            x = x.clone()
+            y = op(x)
+            return y.sum(0)
+
+        x = torch.tensor([0, 1.], requires_grad=True)
+        with self.assertRaisesRegex(
+                RuntimeError,
+                'Getting these operators to work with functionalization requires some extra work'):
+            operator_compile_check(f, (x,), {})
+
+    def test_autograd_registered_at_backend(self, device):
+        lib = self.lib()
+        lib.define("foo(Tensor x) -> Tensor")
+        op = self.ns().foo.default
+
+        class Foo(torch.autograd.Function):
+            @staticmethod
+            def forward(ctx, x):
+                return x.clone()
+
+            @staticmethod
+            def backward(ctx, gx):
+                return gx * 0.5
+
+        lib.impl("foo", Foo.apply, "CPU")
+        lib.impl("foo", Foo.apply, "CUDA")
+        lib.impl("foo", lambda x: x.clone(), "Meta")
+
+        def f(x):
+            y = op(x)
+            return x + y
+
+        x = torch.randn([], requires_grad=True)
+
+        with self.assertRaisesRegex(AssertionError, 'incorrect gradients'):
+            operator_compile_check(f, (x,), {})
+
+        # I'm not sure why this is necessary
+        del lib
+
+    def test_global_state_mutation(self, device):
+        lib = self.lib()
+        lib.define("foo(Tensor x) -> Tensor")
+        op = self.ns().foo.default
+
+        class Foo(torch.autograd.Function):
+            invoked = 0
+
+            @staticmethod
+            def forward(ctx, x):
+                Foo.invoked += 1
+                return x.clone() * Foo.invoked
+
+            @staticmethod
+            def backward(ctx, gx):
+                return gx
+
+        lib.impl("foo", Foo.apply, "CompositeImplicitAutograd")
+
+        def f(x):
+            return op(x)
+
+        x = torch.tensor(3.14159 / 3, requires_grad=True)
+        with self.assertRaisesRegex(AssertionError, "not completely traceable"):
```

Comment:
If dynamo is able to guard on this and trigger a recompile, is that technically considered "traceable" in terms of torch.compile? 

I think the error message is already pretty clear that we're talking about make_fx here and that it is due to depending global state, but maybe its also worth explaining that even though you failed this test, your function may still work with torch.compile?

Reply:
Yes, I'll make it clearer. Ideally what the user provides to operator_compile_check is a single operator (instead of an operator + some more calls).